code/datums/components/basic_mob_attack_telegraph.dm

```diff
@@ -0,0 +1,82 @@
+/**
+ * Delays outgoing attacks which are directed at mobs to give players time to get out of the way
+ */
+/datum/component/basic_mob_attack_telegraph
+	/// Time to wait before attack can complete
+	var/telegraph_duration
+	/// Overlay which we display over targets
+	var/mutable_appearance/target_overlay
+	/// Our current target, if we have one
+	var/mob/living/current_target
+	/// Callback executed when we start aiming at something
+	var/datum/callback/on_began_forecast
+
+/datum/component/basic_mob_attack_telegraph/Initialize(
+	telegraph_icon = 'icons/mob/telegraphing/telegraph.dmi',
+	telegraph_state = ATTACK_EFFECT_BITE,
+	telegraph_duration = 0.3 SECONDS,
+	datum/callback/on_began_forecast,
+)
+	. = ..()
+	if (!isbasicmob(parent))
+		return ELEMENT_INCOMPATIBLE
+
+	target_overlay = mutable_appearance(telegraph_icon, telegraph_state)
+	src.telegraph_duration = telegraph_duration
+	src.on_began_forecast = on_began_forecast
+
+/datum/component/basic_mob_attack_telegraph/RegisterWithParent()
+	. = ..()
+	RegisterSignal(parent, COMSIG_HOSTILE_PRE_ATTACKINGTARGET, PROC_REF(on_attack))
+
+/datum/component/basic_mob_attack_telegraph/UnregisterFromParent()
+	if (current_target)
+		forget_target(current_target)
+	QDEL_NULL(target_overlay)
+	REMOVE_TRAIT(parent, TRAIT_BASIC_ATTACK_FORECAST, REF(src))
+	UnregisterSignal(parent, COMSIG_HOSTILE_PRE_ATTACKINGTARGET)
+	return ..()
+
+/// When we attempt to attack, check if it is allowed
+/datum/component/basic_mob_attack_telegraph/proc/on_attack(mob/living/basic/source, atom/target)
+	SIGNAL_HANDLER
+	if (!isliving(target))
+		return
+	if (HAS_TRAIT(source, TRAIT_BASIC_ATTACK_FORECAST))
```

Comment:
```suggestion
	if (HAS_TRAIT_FROM(source, TRAIT_BASIC_ATTACK_FORECAST, REF(src)))
```

odd feeling that this would pass and do an early return for no reason in case this trait gets used in other things for some reason. might be an unwarranted concern but it feels useful.